src/devices/src/virtio/vsock/event_handler.rs

```diff
@@ -22,201 +22,198 @@
+        if event_set != EventSet::IN {
+            warn!("vsock: rxq unexpected event {:?}", event_set);
+            return false;
         }
 
-        have_used
+        let mut raise_irq = false;
+        if let Err(e) = self.queue_events[RXQ_INDEX].read() {
+            error!("Failed to get vsock rx queue event: {:?}", e);
+        } else if self.backend.has_pending_rx() {
+            raise_irq |= self.process_rx();
+        }
+        raise_irq
     }
 
-    /// Walk the driver-provided TX queue buffers, package them up as vsock packets, and send them to
-    /// the backend for processing.
-    fn process_tx(&mut self) -> bool {
-        debug!("vsock: epoll_handler::process_tx()");
-
-        let mut have_used = false;
-
-        while let Some(head) = self.txvq.pop(&self.mem) {
-            let pkt = match VsockPacket::from_tx_virtq_head(&head) {
-                Ok(pkt) => pkt,
-                Err(e) => {
-                    error!("vsock: error reading TX packet: {:?}", e);
-                    have_used = true;
-                    self.txvq.add_used(&self.mem, head.index, 0);
-                    continue;
-                }
-            };
-
-            if self.backend.send_pkt(&pkt).is_err() {
-                self.txvq.undo_pop();
-                break;
+    fn handle_txq_event(&mut self, event: EpollEvent) -> bool {
+        debug!("vsock: TX queue event");
+
+        let event_set = event.event_set();
+        if event_set != EventSet::IN {
+            warn!("vsock: txq unexpected event {:?}", event_set);
+            return false;
+        }
+
+        let mut raise_irq = false;
+        if let Err(e) = self.queue_events[TXQ_INDEX].read() {
+            error!("Failed to get vsock tx queue event: {:?}", e);
+        } else {
+            raise_irq |= self.process_tx();
+            // The backend may have queued up responses to the packets we sent during
+            // TX queue processing. If that happened, we need to fetch those responses
+            // and place them into RX buffers.
+            if self.backend.has_pending_rx() {
+                raise_irq |= self.process_rx();
             }
+        }
+        raise_irq
+    }
+
+    fn handle_evq_event(&mut self, event: EpollEvent) -> bool {
+        debug!("vsock: event queue event");
 
-            have_used = true;
-            self.txvq.add_used(&self.mem, head.index, 0);
+        let event_set = event.event_set();
+        if event_set != EventSet::IN {
+            warn!("vsock: evq unexpected event {:?}", event_set);
+            return false;
         }
 
-        have_used
+        if let Err(e) = self.queue_events[EVQ_INDEX].read() {
+            error!("Failed to consume vsock evq event: {:?}", e);
+        }
+        false
+    }
+
+    // Helper function to invoke the backend notify() method from different EventHandler
+    // contexts (handler_read, handle_write, etc). Will not be necessary anymore after
+    // switching to the new polly.
```

Comment:
What do you mean by new polly?

Reply:
This was a leftover comment, and I've removed it.